circuit_knitting/cutting/cutting_reconstruction.py

```diff
@@ -75,40 +75,62 @@ def reconstruct_expectation_values(
         ValueError: ``observables`` and ``results`` are of incompatible types.
         ValueError: An input observable has a phase not equal to 1.
     """
-    if isinstance(observables, PauliList) and not isinstance(
-        results, (SamplerResult, PrimitiveResult)
-    ):
-        raise ValueError(
-            "If observables is a PauliList, results must be a SamplerResult or PrimitiveResult instance."
-        )
-    if isinstance(observables, dict) and not isinstance(results, dict):
-        raise ValueError(
-            "If observables is a dictionary, results must also be a dictionary."
-        )
-
-    # If circuit was not separated, transform input data structures to dictionary format
+    # If circuit was not separated, transform input data structures to
+    # dictionary format.  Perform some input validation in either case.
     if isinstance(observables, PauliList):
+        if not isinstance(results, (SamplerResult, PrimitiveResult)):
+            raise ValueError(
+                "If observables is a PauliList, results must be a SamplerResult or PrimitiveResult instance."
+            )
         if any(obs.phase != 0 for obs in observables):
             raise ValueError("An input observable has a phase not equal to 1.")
-        subobservables_by_subsystem = decompose_observables(
-            observables, "A" * len(observables[0])
+        subobservables_by_subsystem: Mapping[Hashable, PauliList] = (
+            decompose_observables(observables, "A" * len(observables[0]))
         )
-        results_dict: dict[Hashable, SamplerResult | PrimitiveResult] = {"A": results}
+        results_dict: Mapping[Hashable, SamplerResult | PrimitiveResult] = {
+            "A": results
+        }
         expvals = np.zeros(len(observables))
 
-    else:
+    elif isinstance(observables, Mapping):
+        if not isinstance(results, Mapping):
+            raise ValueError(
+                "If observables is a dictionary, results must also be a dictionary."
+            )
+        if observables.keys() != results.keys():
```

Comment:
This came about because utility-scale workloads may be run in chunks, saved to disk, and recombined, which means there could very easily be errors during that recombination?

Reply:
That is indeed one motivation.  Mostly I just wanted to add this as an additional sanity check when passing over this code, since the below lines implicitly make this assumption.